### PR TITLE
User should be able to see their own subadmin groups

### LINKED ISSUE #658: Calling `getUserSubadminGroups` with OC10 backend gives 401 Unauthorised
When performing this API call in the phoenix frontend, for a user who is a member of the `admin` group the API call is successful (whether or not they are subadmins of any groups or not) and the correct data payload is returned.  If, however the user is not a member of the specific `admin` group (we call this a global admin and is only given to a few people in our instance) we receive a 401 unauthorised response.

We have no current workaround and are unable to use the subadmin feature for users who we want to give some control over managing groups in Owncloud until this is fixed.

Here is the data payload returned from this failure case:
```
<ocs>
<meta>
<status>failure</status>
<statuscode>997</statuscode>
<message>Unauthorised</message>
</meta>
<data/>
</ocs>
```

Below is a sample screenshot, the call is triggered after login is successful.  It is easy to reproduce by adding the user to the `admin` group (success) then removing the user from the `admin` group (failure).

image

I will raise a PR in `owncloud/core` to this effect.

Once the code in the PR is applied, it will correctly return subadmin groups as long as the user has at least one.  There is an existing workaround implemented (for some reason) against OC10 to handle the API returning an error when users have no subadmin groups (which ideally would not throw an error, but return an empty result set):
https://github.com/owncloud/owncloud-sdk/blob/cabb393f699620bfe4dd828029797e5d652a7d53/src/userManagement.js#L287-L300

### PR DESCRIPTION
**Pull request by rpocklin**
## Description
See related issue below.

## Related Issue
- Fixes https://github.com/owncloud/owncloud-sdk/issues/658

## Motivation and Context
When logging in, we want to check what groups a user is a subadmin of, so we know what UI to present them with if they can manage specific groups.

## Changes to Functionality
Before, you had to be an admin to call this API call.  Now, it is implemented in different ways as follows:

a) If you are a regular user and are requesting your own subadmin groups, it will be successful.
b) If you are a regular user and are requesting someone else subadmin groups, it will fail (401).
c) If you are a subadmin and are requesting yourself, or a user who you manage (via groups) it will be successful.
d) If you are a subadmin and are requesting someone else, it will fail (401).
e) If you are an admin and are requesting anyone, it will be successful.

## How Has This Been Tested?
Works as expected, normal users can call the **owncloud-sdk** without an error response (even if they have are not a subadmin of the group) as per the above use cases.  There are now unit and acceptance tests in the PR.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
- [x] Code changes
- [x] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [x] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by phil-davis** [comment]
CI is failing https://drone.owncloud.com/owncloud/core/28176 because this PR is branched from an old version of master that has out-of-date drone CI settings.

@rpocklin please pull a current master and rebase to that.

While doing that, you can add a link to this PR into the changelog file.

**Comment by rpocklin** [comment]
> CI is failing https://drone.owncloud.com/owncloud/core/28176 because this PR is branched from an old version of master that has out-of-date drone CI settings.
> 
> @rpocklin please pull a current master and rebase to that.
> 
> While doing that, you can add a link to this PR into the changelog file.

Thanks for the feedback, done and done :)

**Comment by mmattel** [comment]
@micbar @phil-davis no reviewer was added before. Pls feel free to add one 😃

Related to https://github.com/owncloud/core/pull/38280

**Comment by phil-davis** [comment]
drone CI got a silly error: https://drone.owncloud.com/owncloud/core/28180/156/12
```
Downloading https://github.com/Medium/phantomjs/releases/download/v2.1.1/phantomjs-2.1.1-linux-x86_64.tar.bz2
Saving to /tmp/phantomjs/phantomjs-2.1.1-linux-x86_64.tar.bz2
Receiving...

Error making request.
Error: connect ETIMEDOUT 52.217.17.156:443
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1191:14)
```

I have restarted CI. I expect that it will pass.

**Comment by mmattel** [comment]
https://drone.owncloud.com/owncloud/core/28182 --> ERR_TUNNEL_CONNECTION_FAILED

**Comment by phil-davis** [comment]
> https://drone.owncloud.com/owncloud/core/28182 --> ERR_TUNNEL_CONNECTION_FAILED

It looks green to me. Maybe the drone server had a "hiccup" rendering the results for a second.

**Comment by rpocklin** [comment]
After further testing I have updated this PR to be more complete, so that it correctly handles admin, subadmin and non-admin users calling this API in the correct manner (for themselves, or for others).  I have tested manually but would recommend it have unit or acceptance tests also but I am unable to do this currently due to lack of experience in PHP and time constraints.

**Review by phil-davis — COMMENTED** [review]
There are code-style things to fix https://drone.owncloud.com/owncloud/core/28185/2/7
It would be good to sort those out, then the rest of CI will run and we will see how the various existing API acceptance tests go.

**Comment by phil-davis** [comment]
https://drone.owncloud.com/owncloud/core/28187/2/7 code style is still failing. It needs to get past that so the real tests will run.

**Comment by mmattel** [comment]
@rpocklin to fix the code styling tests, run `./make test-php-style-fix` from the root of the repo

**Comment by rpocklin** [comment]
Thanks for the instructions guys, will fix this soon.

**Comment by phil-davis** [comment]
Some different status codes get returned in unit tests: https://drone.owncloud.com/owncloud/core/28208/9/8
```
There were 4 failures:

1) OCA\Provisioning_API\Tests\UsersTest::testAddToGroupSuccessfulAsSubadmin
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
 OC\OCS\Result Object (
     'data' => Array ()
     'message' => null
-    'statusCode' => 104
+    'statusCode' => 100
     'items' => null
     'perPage' => null
     'headers' => Array ()
 )

/drone/src/apps/provisioning_api/tests/UsersTest.php:1949

2) OCA\Provisioning_API\Tests\UsersTest::testGetUserSubAdminGroupsNotExistingTargetUser
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
 OC\OCS\Result Object (
     'data' => Array ()
-    'message' => 'User does not exist'
-    'statusCode' => 101
+    'message' => null
+    'statusCode' => 997
     'items' => null
     'perPage' => null
     'headers' => Array ()
 )

/drone/src/apps/provisioning_api/tests/UsersTest.php:2517

3) OCA\Provisioning_API\Tests\UsersTest::testGetUserSubAdminGroupsWithGroups
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
 OC\OCS\Result Object (
     'data' => Array (
-        0 => 'TargetGroup'
     )
     'message' => null
-    'statusCode' => 100
+    'statusCode' => 997
     'items' => null
     'perPage' => null
     'headers' => Array ()
 )

/drone/src/apps/provisioning_api/tests/UsersTest.php:2545

4) OCA\Provisioning_API\Tests\UsersTest::testGetUserSubAdminGroupsWithoutGroups
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
 OC\OCS\Result Object (
     'data' => Array ()
-    'message' => 'Unknown error occurred'
-    'statusCode' => 102
+    'message' => null
+    'statusCode' => 997
     'items' => null
     'perPage' => null
     'headers' => Array ()
 )

/drone/src/apps/provisioning_api/tests/UsersTest.php:2568
```

That needs to be sorted out. It looks to me like more `997` is happening now than should be. I am expecting that only a few test cases will need to change, and those changes should be that a subadmin can "see" their group better than before.

**Comment by phil-davis** [comment]
https://drone.owncloud.com/owncloud/core/28208/59/13 API acceptance test fails:
```
  Scenario: a subadmin cannot add users to groups the subadmin is responsible for                                 # /drone/src/tests/acceptance/features/apiProvisioningGroups-v1/addToGroup.feature:133
    Given these users have been created with default attributes and skeleton files:                               # FeatureContext::theseUsersHaveBeenCreated()
      | username       |
      | brand-new-user |
      | subadmin       |
    And group "brand-new-group" has been created                                                                  # FeatureContext::groupHasBeenCreated()
    And user "subadmin" has been made a subadmin of group "brand-new-group"                                       # FeatureContext::userHasBeenMadeSubadminOfGroup()
    When user "subadmin" tries to add user "brand-new-user" to group "brand-new-group" using the provisioning API # FeatureContext::userTriesToAddUserToGroupUsingTheProvisioningApi()
    Then the OCS status code should be "104"                                                                      # OCSContext::theOCSStatusCodeShouldBe()
      OCS status code is not the expected value 104 got 100
      Failed asserting that two strings are equal.
      --- Expected
      +++ Actual
      @@ @@
      -'104'
      +'100'
    And the HTTP status code should be "200"                                                                      # FeatureContext::thenTheHTTPStatusCodeShouldBe()
    And user "brand-new-user" should not belong to group "brand-new-group"                                        # FeatureContext::userShouldNotBelongToGroup()
```

The sub-admin can now add a member to the group. That ability was removed in PR #31337 back in 2018, because it allows a sub-admin to add any user to "their" group and then change the user's password (thus being able to access the user's data) or delete the user (the sub-admin could delete every other user on the system, leaving just 1 real admin and themselves!)
Also see my comment https://github.com/owncloud/core/issues/38279#issuecomment-757685626

This PR currently does more than its current title: "User should be able to see their own subadmin groups"

IMO we should just fix the "see their own subadmin groups" thing, get that done, and we can have a separate discussion about changing what a sub-admin can and cannot do.

**Comment by rpocklin** [comment]
> https://drone.owncloud.com/owncloud/core/28208/59/13 API acceptance test fails:
> 
> ```
>   Scenario: a subadmin cannot add users to groups the subadmin is responsible for                                 # /drone/src/tests/acceptance/features/apiProvisioningGroups-v1/addToGroup.feature:133
>     Given these users have been created with default attributes and skeleton files:                               # FeatureContext::theseUsersHaveBeenCreated()
>       | username       |
>       | brand-new-user |
>       | subadmin       |
>     And group "brand-new-group" has been created                                                                  # FeatureContext::groupHasBeenCreated()
>     And user "subadmin" has been made a subadmin of group "brand-new-group"                                       # FeatureContext::userHasBeenMadeSubadminOfGroup()
>     When user "subadmin" tries to add user "brand-new-user" to group "brand-new-group" using the provisioning API # FeatureContext::userTriesToAddUserToGroupUsingTheProvisioningApi()
>     Then the OCS status code should be "104"                                                                      # OCSContext::theOCSStatusCodeShouldBe()
>       OCS status code is not the expected value 104 got 100
>       Failed asserting that two strings are equal.
>       --- Expected
>       +++ Actual
>       @@ @@
>       -'104'
>       +'100'
>     And the HTTP status code should be "200"                                                                      # FeatureContext::thenTheHTTPStatusCodeShouldBe()
>     And user "brand-new-user" should not belong to group "brand-new-group"                                        # FeatureContext::userShouldNotBelongToGroup()
> ```
> 
> The sub-admin can now add a member to the group. That ability was removed in PR #31337 back in 2018, because it allows a sub-admin to add any user to "their" group and then change the user's password (thus being able to access the user's data) or delete the user (the sub-admin could delete every other user on the system, leaving just 1 real admin and themselves!)
> Also see my comment [#38279 (comment)](https://github.com/owncloud/core/issues/38279#issuecomment-757685626)
> 
> This PR currently does more than its current title: "User should be able to see their own subadmin groups"
> 
> IMO we should just fix the "see their own subadmin groups" thing, get that done, and we can have a separate discussion about changing what a sub-admin can and cannot do.

Apologies, the change in `addToGroup` was inadvertently added from another current PR I am working on, I will remove it so we can focus on `getUserSubAdminGroups` as that is the intent.

**Comment by rpocklin** [comment]
@phil-davis I am attempting to investigate the failures (which are less now) but phpunit tests keep freezing at 48% on my Ubuntu, are you able to offer suggestions?  I am not familiar with PHP or their toolkits, and i'm using mariadb locally.

```robertp@APL-RP-LT1:~/p/core$ make test-php-unit --debug
GNU Make 4.2.1
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
Updating makefiles....
Updating goal targets....
 File 'test-php-unit' does not exist.
Must remake target 'test-php-unit'.
PHPUNIT="/home/robertp/p/core/lib/composer/phpunit/phpunit/phpunit" build/autotest.sh mysql
Using PHP executable /usr/bin/php
Parsing all files in lib/public for the presence of @since or @deprecated on each method...

Using database oc_autotest
Setup environment for mysql testing on local storage ...
Updated 0 paths from the index
Installing ....
ownCloud was successfully installed
Testing with mysql ...
/home/robertp/p/core/lib/composer/phpunit/phpunit/phpunit --configuration phpunit-autotest.xml --coverage-clover autotest-clover-mysql.xml --coverage-html coverage-html-mysql --log-junit autotest-results-mysql.xml
PHPUnit 8.5.14 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.4.3
Configuration: /home/robertp/p/core/tests/phpunit-autotest.xml
Error:         No code coverage driver is available

...........................................................    59 / 10415 (  0%)
..........................S................................   118 / 10415 (  1%)
...........................................................   177 / 10415 (  1%)
...........................................................   236 / 10415 (  2%)
...........................................................   295 / 10415 (  2%)
...........................................................   354 / 10415 (  3%)
...........................................................   413 / 10415 (  3%)
...........................................................   472 / 10415 (  4%)
...........................................................   531 / 10415 (  5%)
...........................................................   590 / 10415 (  5%)
...........................................................   649 / 10415 (  6%)
...........................................................   708 / 10415 (  6%)
...........................................................   767 / 10415 (  7%)
...........................................................   826 / 10415 (  7%)
...........................................................   885 / 10415 (  8%)
...........................................................   944 / 10415 (  9%)
...........................................................  1003 / 10415 (  9%)
...........................................................  1062 / 10415 ( 10%)
...........................................................  1121 / 10415 ( 10%)
...........................................................  1180 / 10415 ( 11%)
...........................................................  1239 / 10415 ( 11%)
...........................................SS..............  1298 / 10415 ( 12%)
...........................................................  1357 / 10415 ( 13%)
...........................................................  1416 / 10415 ( 13%)
...........................................................  1475 / 10415 ( 14%)
...........................................................  1534 / 10415 ( 14%)
...........................................................  1593 / 10415 ( 15%)
...........................................................  1652 / 10415 ( 15%)
...........................................................  1711 / 10415 ( 16%)
...........................................................  1770 / 10415 ( 16%)
...........................................................  1829 / 10415 ( 17%)
...........................................................  1888 / 10415 ( 18%)
...........................................................  1947 / 10415 ( 18%)
...........................................................  2006 / 10415 ( 19%)
...........................................................  2065 / 10415 ( 19%)
...........................................................  2124 / 10415 ( 20%)
...........................................................  2183 / 10415 ( 20%)
...........................................................  2242 / 10415 ( 21%)
...........................................................  2301 / 10415 ( 22%)
...........................................................  2360 / 10415 ( 22%)
...........................................................  2419 / 10415 ( 23%)
...........................................................  2478 / 10415 ( 23%)
...........................................................  2537 / 10415 ( 24%)
...........................................................  2596 / 10415 ( 24%)
...........................................................  2655 / 10415 ( 25%)
...........................................................  2714 / 10415 ( 26%)
...........................................................  2773 / 10415 ( 26%)
...........................................................  2832 / 10415 ( 27%)
...........................................................  2891 / 10415 ( 27%)
...........................................................  2950 / 10415 ( 28%)
..............................................S............  3009 / 10415 ( 28%)
...........................................................  3068 / 10415 ( 29%)
...........................................................  3127 / 10415 ( 30%)
..................S........................................  3186 / 10415 ( 30%)
...................................................S.......  3245 / 10415 ( 31%)
...........................................................  3304 / 10415 ( 31%)
..................................S........................  3363 / 10415 ( 32%)
...........................................................  3422 / 10415 ( 32%)
................S..........................................  3481 / 10415 ( 33%)
..............................................S............  3540 / 10415 ( 33%)
...........................................................  3599 / 10415 ( 34%)
...........................................................  3658 / 10415 ( 35%)
...........................................................  3717 / 10415 ( 35%)
...........................................................  3776 / 10415 ( 36%)
...........................................................  3835 / 10415 ( 36%)
...........................................................  3894 / 10415 ( 37%)
...........................................................  3953 / 10415 ( 37%)
...........................................................  4012 / 10415 ( 38%)
...........................................................  4071 / 10415 ( 39%)
...........................................................  4130 / 10415 ( 39%)
...........................................................  4189 / 10415 ( 40%)
...........................................................  4248 / 10415 ( 40%)
.....................................................FF....  4307 / 10415 ( 41%)
...........................................................  4366 / 10415 ( 41%)
...........................................................  4425 / 10415 ( 42%)
...........................................................  4484 / 10415 ( 43%)
...........................................................  4543 / 10415 ( 43%)
...........................................................  4602 / 10415 ( 44%)
...........................................................  4661 / 10415 ( 44%)
...........................................................  4720 / 10415 ( 45%)
...............SSSSSSSSSSSSSSSSSS..........................  4779 / 10415 ( 45%)
..SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.....................  4838 / 10415 ( 46%)
...........................................................  4897 / 10415 ( 47%)
...........................................................  4956 / 10415 ( 47%)
...........................................................  5015 / 10415 ( 48%)
...........................FFFF........```

**Comment by phil-davis** [comment]
I know that some of the unit tests somehow use a lot of memory. You can run just the test(s) that you know are a probledm, get those passing, and let CI run the full suite. Something like:
```
make test-php-unit NOCOVERAGE=true TEST_PHP_SUITE=apps/provisioning_api/tests/UsersTest.php
```

**Comment by rpocklin** [comment]
> I know that some of the unit tests somehow use a lot of memory. You can run just the test(s) that you know are a probledm, get those passing, and let CI run the full suite. Something like:
> 
> ```
> make test-php-unit NOCOVERAGE=true TEST_PHP_SUITE=apps/provisioning_api/tests/UsersTest.php
> ```

Thanks that helped me move forward cheers.

**Comment by rpocklin** [comment]
I keep getting the following error running the **test-acceptance-api** tests locally, any ideas?

```
Nothing to install or update
Generating autoload files
composer/package-versions-deprecated: Generating version class...
composer/package-versions-deprecated: ...done generating version class
61 packages you are using are looking for funding.
Use the `composer fund` command to find out more!
./tests/acceptance/run.sh --type api
Script path: /home/robertp/p/core/tests/acceptance
Using php inbuilt server for running scenario ...
8080
6151
8180
6152
[Tue Feb 23 17:12:15 2021] PHP 7.4.3 Development Server (http://localhost:8180) started
[Tue Feb 23 17:12:15 2021] PHP 7.4.3 Development Server (http://localhost:8080) started
[Tue Feb 23 17:12:20 2021] 127.0.0.1:38312 Accepted
[Tue Feb 23 17:12:20 2021] 127.0.0.1:38312 [302]: POST /ocs/v2.php/apps/testing/api/v1/occ
[Tue Feb 23 17:12:20 2021] 127.0.0.1:38312 Closing
-:2: parser error : Start tag expected, '<' not found

^
[Tue Feb 23 17:12:20 2021] 127.0.0.1:38314 Accepted
[Tue Feb 23 17:12:20 2021] 127.0.0.1:38314 [302]: POST /ocs/v2.php/apps/testing/api/v1/occ
[Tue Feb 23 17:12:20 2021] 127.0.0.1:38314 Closing
-:2: parser error : Start tag expected, '<' not found

^
Could not set mail_domain on http://localhost:8080/ocs/v2.php/apps/testing/api/v1/occ. Result:
''
[Tue Feb 23 17:12:20 2021] 127.0.0.1:38316 Accepted
[Tue Feb 23 17:12:20 2021] 127.0.0.1:38316 [302]: POST /ocs/v2.php/apps/testing/api/v1/occ
[Tue Feb 23 17:12:20 2021] 127.0.0.1:38316 Closing
-:2: parser error : Start tag expected, '<' not found

^
runsh: Exit code of main run:
make: *** [Makefile:207: test-acceptance-api] Error 1```

**Comment by phil-davis** [comment]
You need to have a local ownCloud install that will already work properly on localhost:8080 (for example)
I find it easiest to already have my system-under-test running before starting the tests. Then I can login to it manually on the web UI and know that it is working. 

Sometimes I get this problem because I have enable and app and it wants `php occ upgrade` to be done.

To run the tests, I then do:
```
export TEST_SERVER_URL=http://localhost:8080
```

And then run a particular test(s):
```
make test-acceptance-api BEHAT_FEATURE=tests/acceptance/features/apiCapabilities/capabilities.feature
```

If `TEST_SERVER_URL` then the test runner script does not start up its own PHP dev server.

**Comment by rpocklin** [comment]
Thanks @phil-davis I believe it all passes now =)  Ssomeone added in the acceptance tests relating to this *@issue-owncloud-sdk-658* for me, which was helpful and thoughtful.

**Comment by phil-davis** [comment]
The code looked like it will now allow an ordinary user who is not a subadmin of any groups to query about what groups they are sub-admin of, so I added an acceptance test scenario for that. They get "success" OCS and HTTP status with an empty list - which seems correct.

**Review by phil-davis — APPROVED** [review]
LGTM - also needs review from a developer.

**Review by VicDeo — APPROVED** [review]

File: apps/provisioning_api/lib/Users.php [hunk 1]
```diff
@@ -628,22 +628,39 @@ public function removeSubAdmin($parameters) {
...
+		}
+
+		$currentLoggedInUserId = $currentLoggedInUser->getUID();
+
+		$targetUserId = $parameters['userid'];
+		$targetUser = $this->userManager->get($targetUserId);
+
 		// Check if the user exists
-		if ($user === null) {
+		if ($targetUser === null) {
```
Comment (JanAckermann) [inline apps/provisioning_api/lib/Users.php:644]: Needs to investigated, I am Not quite sure, but isn't it against our security policy, to Show If a User exists, to a

Comment (VicDeo) [reply apps/provisioning_api/lib/Users.php:644]: This requires a logged in user... I think  it's better to use a share dialog if you need some usernames

Comment (phil-davis) [reply apps/provisioning_api/lib/Users.php:644]: @JanAckermann "to a..." what?

**Review by AlexAndBear — APPROVED** [review]

File: apps/provisioning_api/lib/Users.php [hunk 2]
```diff
@@ -628,22 +628,39 @@ public function removeSubAdmin($parameters) {
...
 			return new Result(null, 101, 'User does not exist');
```
Comment (JanAckermann) [inline apps/provisioning_api/lib/Users.php:645]:
```suggestion
			return new Result(null, API::RESPOND_NOT_FOUND, 'The requested user could not be found');
```

For consistency purposes

**Review by AlexAndBear — CHANGES_REQUESTED** [review]

**Comment by rpocklin** [comment]
> The code looked like it will now allow an ordinary user who is not a subadmin of any groups to query about what groups they are sub-admin of, so I added an acceptance test scenario for that. They get "success" OCS and HTTP status with an empty list - which seems correct.

Yep that's right, thanks for adding that.

**Comment by phil-davis** [comment]
@JanAckermann IMO this is ready. Please review again.

**Comment by JanAckermann** [comment]
Unit Tests are failing

**Comment by rpocklin** [comment]
Updated the tests, everything passes now.  Incorporating the change @JanAckermann suggested meant I needed to update the `tests/acceptance/features/apiProvisioningGroups-v2/getSubAdminGroups.feature` file to change the OCS and HTTP response code from `400` to `404.

**Review by AlexAndBear — APPROVED** [review]

**Review by phil-davis — APPROVED** [review]

**Pull request merged**